File: edukate-notifier/src/main/java/io/github/sanyavertolet/edukate/notifier/entities/CheckedNotification.java
package io.github.sanyavertolet.edukate.notifier.entities;

import com.fasterxml.jackson.annotation.JsonTypeName;
import io.github.sanyavertolet.edukate.common.checks.CheckStatus;
import io.github.sanyavertolet.edukate.common.notifications.CheckedNotificationCreateRequest;
import io.github.sanyavertolet.edukate.notifier.dtos.CheckedNotificationDto;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@JsonTypeName("checked")
public final class CheckedNotification extends BaseNotification {
    private String submissionId;
    private String problemId;
    private CheckStatus status;

    public CheckedNotification(
            String uuid, String targetUserId, String submissionId, String problemId, CheckStatus status
    ) {
        super(uuid, targetUserId);
        this.submissionId = submissionId;
        this.problemId = problemId;
        this.status = status;
    }

    @Override
    public CheckedNotificationDto toDto() {
        return new CheckedNotificationDto(getUuid(), getIsRead(), getCreatedAt(), submissionId, problemId, status);
    }

    public static CheckedNotification fromCreationRequest(CheckedNotificationCreateRequest creationRequest) {
        return new CheckedNotification(
                creationRequest.getUuid(), creationRequest.getTargetUserId(),
                creationRequest.getSubmissionId(), creationRequest.getProblemId(), creationRequest.getStatus()
        );
    }
}

File: edukate-notifier/src/main/kotlin/io/github/sanyavertolet/edukate/notifier/entities/CheckedNotification.kt
package io.github.sanyavertolet.edukate.notifier.entities

import com.fasterxml.jackson.annotation.JsonTypeName
import io.github.sanyavertolet.edukate.common.checks.CheckStatus
import io.github.sanyavertolet.edukate.common.notifications.CheckedNotificationCreateRequest
import io.github.sanyavertolet.edukate.notifier.dtos.CheckedNotificationDto
import java.time.Instant
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.Indexed

@JsonTypeName("checked")
data class CheckedNotification(
    @field:Id override val id: String? = null,
    @field:Indexed(unique = true) override val uuid: String,
    override val targetUserId: String,
    override val isRead: Boolean = false,
    @field:CreatedDate override val createdAt: Instant? = null,
    val submissionId: String,
    val problemId: String,
    val status: CheckStatus,
) : BaseNotification() {
    override fun toDto() =
        CheckedNotificationDto(
            uuid = uuid,
            isRead = isRead,
            createdAt = requireNotNull(createdAt) { "createdAt is null for notification uuid=$uuid" },
            submissionId = submissionId,
            problemId = problemId,
            status = status,
        )

    override fun markAsRead() = copy(isRead = true)

    companion object {
        @JvmStatic
        fun fromCreationRequest(creationRequest: CheckedNotificationCreateRequest) =
            CheckedNotification(
                uuid = creationRequest.uuid,
                targetUserId = creationRequest.targetUserId,
                submissionId = creationRequest.submissionId,
                problemId = creationRequest.problemId,
                status = creationRequest.status,
            )
    }
}
